### PR TITLE
US81171 - Ctwomey/view updates data

### PR DESCRIPTION
**Pull request by ctwomey1**
behavior for getting updates aggregate number.  Branched off #266 

US81171.

Acceptance criteria:
- Pulls updates data per course on initial load
- search provides correct updates when tile renders
- pin/unpin actions shouldn't impact the update number shown
- configuration from lms influences update count.

An update is defined by the configuration found in the lms:

screen shot 2017-01-20 at 4 00 00 pm

Each can be achieved by impersonating a student enrolled in a course and doing the relevant action (ie do a quiz, submit an assignment), or by impersonating a teacher and doing the relevant action (grade/reply to post, etc).

### CODE REVIEW

File: d2l-course-tile.html [hunk 1]
```diff
@@ -294,6 +309,34 @@
 						break;
 				}
 			},
+			setCourseUpdates: function(updates) {
+				if (updates > 99) {
+					this._courseUpdates = '99+';
+					this._hasCourseUpdates = true;
+					this.toggleClass('d2l-updates-hidden', false, this.$.courseUpdates);
+				} else if (updates > 0 && updates <= 99) {
```
Comment (apalaniuk) [inline d2l-course-tile.html]: Could drop the `<= 99` check (but probably not worth updating right now)

File: d2l-utility-behavior.html [hunk 2]
```diff
@@ -66,6 +66,12 @@
 			var selfLink = entity.getLinkByRel('self');
 			return selfLink.href;
 		},
+		getOrgUnitId: function(enrollmentId) {
+			if (enrollmentId && typeof String) {
```
Comment (ctwomey1) [inline d2l-utility-behavior.html]: Ha def a finger slip on that one, good catch

**Comment by ctwomey1** [comment]
Putting back into WIP based on some feedback about configuration of the updates number.

File: d2l-my-courses.html [hunk 3]
```diff
@@ -111,7 +113,8 @@
 			my-enrollments-entity="[[_lastEnrollmentsSearchResponse]]"
 			lastEnrollmentsSearchResponse="[[_lastEnrollmentsSearchResponse]]"
 			locale="[[locale]]"
-      		show-course-code="[[showCourseCode]]">
+      		show-course-code="[[showCourseCode]]"
+			update-count-map="[[updateCountMap]]">
```
Comment (apalaniuk) [inline d2l-my-courses.html]: This should probably be updated with @ryantmer's change to create `d2l-all-courses` dynamically - since `d2l-all-courses` uses `UpdatesBehavior` directly, I don't think there are any issues due to losing binding or anything though.

File: d2l-updates-behavior.html [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+			updateCountMap: {
+				type: Object,
+				value: function() {
+					return {};
+				}
+			},
+			_updatesURI: {
+				type: String,
+				value: function() {
+					return '/d2l/api/le/1.18/updates/myUpdates/?orgUnitIdsCSV=';
```
Comment (apalaniuk) [inline d2l-updates-behavior.html]: Does this need to be unique per element?

File: d2l-updates-behavior.html [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+				}
+			},
+			_ajaxElement: {
+				type: Object,
+				value: function() {
+					return document.createElement('d2l-ajax');
```
Comment (ryantmer) [inline d2l-updates-behavior.html]: This should be created on `getUpdates` I think; it's a pretty heavy thing to have here (also not really needed until the function is called anyway).

File: d2l-all-courses.html [hunk 6]
```diff
@@ -254,11 +257,6 @@ <h2 class="d2l-heading-3">{{localize('unpinnedCourses')}}</h2>
 					size="100">
 				</d2l-loading-spinner>
 			</div>
-
-			<d2l-loading-spinner
```
Comment (ryantmer) [inline d2l-all-courses.html]: Why was this removed? :o We're still using it further down (or at least, we're show/hiding it)

File: d2l-course-tile.html [hunk 7]
```diff
@@ -317,13 +332,41 @@
...
+			_onUpdatesChange: function(updateChange) {
+				if (this.enrollmentId) {
+					var orgUnitId = this.getOrgUnitId(this.enrollmentId);
+					var update = updateChange.base[orgUnitId];
+					if (update) {
+						this.setCourseUpdates(update);
+					}
+				}
+			},
+			_shortCourseText: function() {
```
Comment (ryantmer) [inline d2l-course-tile.html]: This appears to be unused - not sure if it's usage is still WIP. If not, then this and `_hasCourseUpdates` can both go since it's not used anywhere else.

File: d2l-updates-behavior.html [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+			var updates = this._updatesURI + csvOrgId;
+			this._ajaxElement.url = updates;
+			this._ajaxElement.onResponse = this._onUpdatesResponse.bind(this);
+			this._ajaxElement.method = 'GET';
+			this._ajaxElement.generateRequest();
+		},
+		_onUpdatesResponse: function(response) {
+			if (response.detail.status === 200) {
+				response.detail.xhr.response['Objects'].forEach(function(object) {
+					this.set('updateCountMap.' + object['OrgUnitId'], object['UnreadDiscussions'] + object['UnapprovedDiscussions'] +
```
Comment (ryantmer) [inline d2l-updates-behavior.html]: nit: long/awkward line is long/awkward

File: d2l-all-courses.html (hunk 6 above)
Comment (ctwomey1) [inline d2l-all-courses.html]: probably just a bad merge

File: d2l-course-tile.html (hunk 7 above)
Comment (ctwomey1) [inline d2l-course-tile.html]: you're right it can go, it was a holdover from when it was a dom-if

File: d2l-updates-behavior.html (hunk 4 above)
Comment (ctwomey1) [inline d2l-updates-behavior.html]: no, should it be defined inline?

Comment (apalaniuk) [inline d2l-updates-behavior.html]: Yeah, or directly in a `createUpdatesUrl(orgUnitIds)` or similar function so it's still simple to test

**Comment by ryantmer** [comment]
Recent changes look good - I'll do the formal "Approve" once the UI PR has been merged and this one is rebased/merged (just as a last once-over).

File: d2l-course-tile.html [hunk 9]
```diff
@@ -246,7 +251,14 @@
 
 				var organizationLink = this.enrollment.getLinkByRel(/\/organization$/);
 				this._organizationUrl = organizationLink.href + '?embedDepth=1';
-
+				//because the tile might not be init before the first search we need to check the map on ready
```
Comment (ryantmer) [inline d2l-course-tile.html]: Can this be delayed? Just to keep `ready` fairly lightweight (also if it's deferred until later I don't know that you'd need the `if (this.updateCountMap)` check anymore... maybe?). I'm thinking maybe move it into `_onOrganizationResponse` - I realize it's not actually related to the organization response coming back, but that's the point where the tile is "ready" from a human-usable perspective, so it seems reasonable to do it in there.

File: d2l-course-tile.html [hunk 10]
```diff
@@ -330,25 +342,33 @@
...
-					this._hasCourseUpdates = true;
 					this.toggleClass('d2l-updates-hidden', false, this.$.courseUpdates);
 				} else {
 					this._courseUpdates = 0;
-					this._hasCourseUpdates = false;
 					this.toggleClass('d2l-updates-hidden', true, this.$.courseUpdates);
 				}
 			},
+			_onUpdatesChange: function(updateChange) {
+				if (this.enrollmentId) {
```
Comment (ryantmer) [inline d2l-course-tile.html]: Strictly speaking this check is redundant - you do the same one in `getOrgUnitId` itself.

File: d2l-course-tile.html [hunk 11]
```diff
@@ -330,25 +342,33 @@
...
+			},
+
 			_enrollmentChanged: function() {
 				if (!this.delayLoad && this.enrollment) {
 					var organizationLink = this.enrollment.getLinkByRel(/\/organization$/);
 					if (!this._organizationUrl || this._organizationUrl.indexOf(organizationLink.href) !== 0) {
 						this._organizationUrl = organizationLink.href + '?embedDepth=1';
 						this.$.organizationRequest.generateRequest();
 					}
+
```
Comment (ryantmer) [inline d2l-course-tile.html]: nit: extra space ("while you're here" :P)

File: d2l-updates-behavior.html [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+				type: Object,
+				value: function() {
+					return {};
+				}
+			},
+			//config from outside the widget
+			courseUpdatesConfig: {
+				type: Object
+			},
+			_ajaxElement: {
```
Comment (ryantmer) [inline d2l-updates-behavior.html]: This doesn't actually need to be a Polymer property (not like... a huge deal, but it's not being observed and not in the DOM or anything, so no need).

File: d2l-updates-behavior.html [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+		getUpdates: function(csvOrgId) {
+			if (this.courseUpdatesConfig) {
+				if (typeof this.courseUpdatesConfig === 'string') {
+					this.courseUpdatesConfig = JSON.parse(this.courseUpdatesConfig);
+				}
+				this._ajaxElement = this._ajaxElement || document.createElement('d2l-ajax');
+				var updates = this._generateUpdatesUri(csvOrgId);
+				if (updates) {
+					this._ajaxElement.url = updates;
+					this._ajaxElement.onResponse = this._onUpdatesResponse.bind(this);
```
Comment (ryantmer) [inline d2l-updates-behavior.html]: A perhaps more Polymer-pragmatic way to do this is to use `this.listen`, i.e. `this.listen(this._ajaxElement, 'iron-ajax-response', '_onUpdatesResponse')` (and then `this.unlisten` in `_onUpdatesResponse`).

File: d2l-updates-behavior.html [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+				}
+			}
+		},
+		_generateUpdatesUri: function(csvOrgId) {
+			if (csvOrgId && typeof csvOrgId  === 'string') {
+				return '/d2l/api/le/1.18/updates/myUpdates/?orgUnitIdsCSV=' + csvOrgId;
+			}
+		},
+		_onUpdatesResponse: function(response) {
+			if (response.detail.status === 200) {
```
Comment (ryantmer) [inline d2l-updates-behavior.html]: Does this API ever 304 Not Modified? (I'm assuming not, but might be worth checking)

File: test/index.html [hunk 15]
```diff
@@ -25,7 +25,8 @@
 				'./d2l-touch-menu-item/d2l-touch-menu-item.html',
 				'./d2l-search-widget-custom/d2l-search-widget-custom.html',
 				'./d2l-utility-behavior/d2l-utility-behavior.html',
-				'./localize-behavior/localize-behavior.html'
+				'./localize-behavior/localize-behavior.html',
+				'./d2l-updates-behavior/d2l-updates-behavior.html'
```
Comment (ryantmer) [inline test/index.html]: nit: BUT ALL THE OTHER ONES ARE ALPHABETICAL 😿 😿 😿 😿 😿 😿 y chester y

File: d2l-updates-behavior.html (hunk 14 above)
Comment (ctwomey1) [inline d2l-updates-behavior.html]: http://docs.valence.desire2learn.com/res/updates.html, doesn't appear to

File: d2l-updates-behavior.html (hunk 13 above)
Comment (ctwomey1) [inline d2l-updates-behavior.html]: Care to elaborate on why?  Seems like more overhead for the same result.  We do use this method a few places in this repo currently (doesn't mean it's right, just saying)

File: d2l-updates-behavior.html (hunk 12 above)
Comment (ctwomey1) [inline d2l-updates-behavior.html]: It doesn't need to, but assigning it to polymer gives me scoping for free i think because of the mixin jazz?  Otherwise i'd be attaching it to window.d2l.mycourses.updatesajax, which kind of opens it up to weird abuse i think.

File: d2l-course-tile.html (hunk 9 above)
Comment (ctwomey1) [inline d2l-course-tile.html]: This is a pretty lightweight operation, more or less just setting a property. However i don't believe it can be delayed.

This had to be added because of the way all courses gets init now.  The initial observable event triggered in mycourses happens pre course tiles in d2l-all-courses.

File: test/index.html (hunk 15 above)
Comment (ctwomey1) [inline test/index.html]: some people just like to watch the world burn

File: d2l-updates-behavior.html (hunk 13 above)
Comment (ryantmer) [inline d2l-updates-behavior.html]: Yeah, I'm not super committed to either, just an option that uses more "Polymer-y" things.

File: d2l-course-tile.html (hunk 9 above)
Comment (ryantmer) [inline d2l-course-tile.html]: Truth, it is lightweight - it's not really a performance thing in this case, it's more a "thing doesn't belong here" in a flow sense, as in this isn't required for the element to operate properly. Considering we don't even fetch the organization here (and that is super required for the tile to operate correctly), the updates don't really need to be set this early anyway.

Yarr, the observable event will happen before All Courses is around, but that wouldn't matter in this case - if the event has already happened, that just means All Courses (and therefore this tile) is being created when `updateCountMap` is already populated. If that's the case, `if (this.updateCountMap)` will be true, and we're golden to do it in `_onOrganizationsResponse` - if it still hasn't been set (e.g. if All Courses was opened really quickly, and the Valence updates route is super slow), then then `_onUpdatesChange` will take care of it.

So like:

- Updates call happens, sets `updateCountMap` in d2l-my-courses
- User opens All Courses, `updateCountMap` is propagated to it, then to its child course tiles
- Course tile fetches organization
- Course tile receives organization, sets text/link of the tile, fetches image, and grabs its update count from `updateCountMap`

File: d2l-updates-behavior.html (hunk 12 above)
Comment (ryantmer) [inline d2l-updates-behavior.html]: Oh derp, the fact that this was in a behaviour completely whooshed over my head, thought it was an element. nvm, carry on! :P

File: test/index.html (hunk 15 above)
Comment (ryantmer) [inline test/index.html]: :fire: 🌎

File: d2l-course-tile.html (hunk 9 above)
Comment (ctwomey1) [inline d2l-course-tile.html]: ok, you've convinced me 👍.

My thought process was mostly around I only want to do this once because of the new way we are doing things.  However that probably didn't cover the case of a tile's enrollment changing.  Moving it to the organization response is probably the right thing to do to ensure that the update number is always correct.

Comment (ryantmer) [inline d2l-course-tile.html]: Yeah, I mean it is unlikely that the enrollments won't have been set right away (since we're `dom-repeat`ing these to create them), but it is possible. But yeah, guaranteed we'll be fetching the right update number at that point.

File: d2l-course-tile.html [hunk 16]
```diff
@@ -330,18 +334,22 @@
...
 					this._courseUpdates = updates;
-					this._hasCourseUpdates = true;
 					this.toggleClass('d2l-updates-hidden', false, this.$.courseUpdates);
 				} else {
 					this._courseUpdates = 0;
-					this._hasCourseUpdates = false;
 					this.toggleClass('d2l-updates-hidden', true, this.$.courseUpdates);
 				}
 			},
+			_onUpdatesChange: function(updateChange) {
```
Comment (ryantmer) [inline d2l-course-tile.html]: Huh, only just noticed this - this and what's being done in `_onOrganizationResponse` is the same. Could just use `this.updateCountMap[orgUnitId]` here instead of the passed-in value, and call this from the response handler.

Comment (ctwomey1) [inline d2l-course-tile.html]: what's passed in is guaranteed to be pushed down fully from the event, the map on the object is subject to polymer propagating things fast enough

Comment (ryantmer) [inline d2l-course-tile.html]: Yeah fair. I haven't personally come across problems with that, but that was also generally in simpler cases than a potentially-complex object. Sounds good!

**Comment by JoshuaKlassen** [comment]
Testing notes:

---
Acceptance Criteria: 
- [X] Pulls updates data per course on initial load 

- [ ] Search provides correct updates when tile renders 
 - If canonical namespaces is on, then no number will show up in the All Courses view
- [X] Pin/unpin actions shouldn't impact the number shown 

- [X] Configuration from lms influences update count

---
Notes:
- A new Assignment Submission counts as 4 updates?
- Un-attempted Quizzes is really just how many quizzes have not been completed by everybody, not an issue or anything related to this PR, just confusing

**Comment by ryantmer** [comment]
[1942f66 irl](http://static.fjcdn.com/pictures/Csi_3c4720_195814.jpg)

**Comment by JoshuaKlassen** [comment]
Notification count appears in All Courses view now with and without Canonical Namespaces!

---
Acceptance Criteria:

- [X] Pulls updates data per course on initial load

- [X] Search provides correct updates when tile renders

- [X] Pin/unpin actions shouldn't impact the number shown

- [X] Configuration from lms influences update count
